jolt/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+//! Translation layer for using NexusVM with Jolt (https://github.com/a16z/jolt).
+//!
+//! JoltVM works with a superset of NexusVM instruction set, hence the mapping is almost identical.
+//! Mainly, it's required to modify the link script for the memory shift, fetch additional sections
+//! from the ELF file, and build a trace of memory accesses.
+
+#![allow(clippy::type_complexity)]
+
+use jolt_common::{constants::MEMORY_OPS_PER_INSTRUCTION, rv_trace as jolt_rv};
+use jolt_core::{
+    jolt::vm::{
+        bytecode::BytecodeRow as JoltBytecodeRow,
+        rv32i_vm::{self, RV32ISubtables, C, M, RV32I},
+        Jolt,
+    },
+    poly::{commitment::hyrax::HyraxScheme, field::JoltField},
+    utils::thread::unsafe_allocate_zero_vec,
+};
+
+use rayon::{
+    iter::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIterator},
+    slice::ParallelSliceMut,
+};
+
+const LOG_TARGET: &str = "nexus-jolt";
+
+pub type F = ark_bn254::Fr;
+pub type PCS = HyraxScheme<ark_bn254::G1Projective>;
+pub type JoltPreprocessing = jolt_core::jolt::vm::JoltPreprocessing<F, PCS>;
+pub type JoltProof = jolt_core::jolt::vm::JoltProof<C, M, F, PCS, RV32I, RV32ISubtables<F>>;
+pub type JoltCommitments = jolt_core::jolt::vm::JoltCommitments<PCS>;
+
+mod convert;
+mod error;
+
+pub mod parse;
+pub mod trace;
+
+pub use error::Error;
+
+/// Wrapper for initialized VM.
+pub struct VM {
+    /// Initialized Nexus VM.
+    vm: nexus_riscv::eval::VM,
+
+    /// Instructions section.
+    insts: Vec<jolt_rv::ELFInstruction>,
+
+    /// Flattened initial state of memory.
+    mem_init: Vec<(u64, u8)>,
+}
+
+impl VM {
+    pub fn bytecode_size(&self) -> usize {
+        self.insts.len()
+    }
+}
+
+pub fn preprocess(vm: &VM) -> JoltPreprocessing {
+    rv32i_vm::RV32IJoltVM::preprocess(
+        vm.insts.clone(),
+        vm.mem_init.clone(),
+        1 << 22,
```

Comment:
It would be great if we could have some descriptive comments here for any of the magic numbers that pop up as part of converting the Jolt API. Will make it easier to maintain the interface in the future.

Reply:
+1, extracting to a named constant can also help keep it self-documenting.